### DISCUSSION TITLE
Need advice: lsblk dreams up partitions?

### DISCUSSION THREAD

**Discussion by schildbach** [util-linux/util-linux#2178] [General]
I've got an SSD that I use LUKS2 on, and in that LUKS there is a filesystem (directly, without partition table, as I believe). However, lsblk prints this:

```
# lsblk -p /dev/sdb
NAME                    MAJ:MIN RM   SIZE RO TYPE  MOUNTPOINTS
/dev/sdb                  8:16   0   3,6T  0 disk  
├─/dev/sdb1               8:17   0 240,9G  0 part  
├─/dev/sdb2               8:18   0 690,7G  0 part  
└─/dev/mapper/sdb_crypt 253:0    0   3,6T  0 crypt /mnt/xxx
```

Now I wonder where do `/dev/sdb1` and `/dev/sdb2` parts stem from? And where do their capacities live? As you can see the 3,6T is taken up by `/dev/mapper/sdb_crypt` already, so I'd say it's impossible to fit another 240,9G and 690,7G in.

I have two HDDs in the same system which I created the same way (I believe), but years earlier. They do not exhibit this behavior, e.g.

```
# lsblk -p /dev/sdc
NAME                    MAJ:MIN RM  SIZE RO TYPE  MOUNTPOINTS
/dev/sdc                  8:32   0  5,5T  0 disk  
└─/dev/mapper/sdc_crypt 253:1    0  5,5T  0 crypt /mnt/xxx
```

This is how I'd expect `/dev/sdb` to look like too.

Fdisk doesn't show any partition tables:

```
# fdisk -l /dev/sdb
Disk /dev/sdb: 3,64 TiB, 4000787030016 bytes, 7814037168 sectors
Disk model: xxx 
Units: sectors of 1 * 512 = 512 bytes
Sector size (logical/physical): 512 bytes / 512 bytes
I/O size (minimum/optimal): 512 bytes / 512 bytes
```

Parted doesn't seem to find a partition table either:

```
# parted -l /dev/sdb
…
Error: /dev/sdb: unrecognised disk label
Model: xxx
Disk /dev/sdb: 4001GB
Sector size (logical/physical): 512B/512B
Partition Table: unknown
Disk Flags: 
…
```

So where do the ghost partitions come from? And how do I get rid of them in a safe way?

I have xxx'ed out some irrelevant data.

**Comment by t-8ch** [comment]
> Now I wonder where do /dev/sdb1 and /dev/sdb2 parts stem from?

Afaik these partitions should come from the kernels partition parser itself, not util-linux.

> And where do their capacities live?

Try `lsblk -o +START,PTTYPE,PARTTYPE`.


Are these partitions detected on every boot/hotplug?

**Reply by schildbach** [reply]
> > Now I wonder where do /dev/sdb1 and /dev/sdb2 parts stem from?
> 
> Afaik these partitions should come from the kernels partition parser itself, not util-linux.

Thanks for the hint. I looked ad `dmesg`, and found the line

```
[    2.567079]  sdb: AHDI sdb1 sdb2
```

for the drive that should *not* have partirtions. On the other hand, there is a drive that *has*  a GPT partition table and two partitions, and it yields the line

```
[    2.575979]  sda: sda1 sda2
```

What does "AHDI" mean? Why does sdb have it, and sda not?

> > And where do their capacities live?
> 
> Try `lsblk -o +START,PTTYPE,PARTTYPE`.

The START field doesn't exist for my lsblk. The other fields are empty for the "ghost" partitions:

```
# lsblk -o +PTTYPE,PARTTYPE /dev/sdb
NAME        MAJ:MIN RM   SIZE RO TYPE  MOUNTPOINTS PTTYPE PARTTYPE
sdb           8:16   0   3,6T  0 disk                     
├─sdb1        8:17   0 240,9G  0 part                     
├─sdb2        8:18   0 690,7G  0 part                     
└─sdb_crypt 253:0    0   3,6T  0 crypt /mnt/xxx          
```

On my drive that has partitions, I get:

```
# lsblk -o +PTTYPE,PARTTYPE /dev/sda
NAME   MAJ:MIN RM   SIZE RO TYPE MOUNTPOINTS           PTTYPE PARTTYPE
sda      8:0    0 167,7G  0 disk                       gpt    
├─sda1   8:1    0   487M  0 part /boot/efi             gpt    c12a7328-f81f-11d2-ba4b-00a0c93ec93b
└─sda2   8:2    0 167,2G  0 part /xxx/                 gpt    0fc63daf-8483-4772-8e79-3d69d8477de4
```

> Are these partitions detected on every boot/hotplug?

I would say yes. I rebooted twice since.

**Reply by schildbach** [reply]
> What does "AHDI" mean?

Answering my own question: AHDI is a format used by Atari computers. See my ending post below.

**Comment by karelzak** [comment]
lsblk does not parse anything; it reads all info from /sys. The kernel displays partitions as a subdirectory (e.g. see `ls -d /sys/block/sdb/sd*` and other relations by `/sys/block/*/holders/` and  `/sys/block/*/slaves/` .

**Comment by schildbach** [comment]
I just found references to this bug in the kernel: https://bugzilla.kernel.org/show_bug.cgi?id=211015, https://bugs.launchpad.net/ubuntu/+source/util-linux/+bug/1531404
Apparently, AHDI is a format for Atati computers and their parser misdetects LUKS headers every now and then…

**Comment by schildbach** [comment]
I just worked around my issue by converting my LUKS2 pbkdf from `argon2i` to `argon2id`:

`cryptsetup luksConvertKey /dev/sdb --pbkdf argon2id`

Apparently that changed the LUKS header enough so that Atari is no longer confused.

I'm closing this, thanks for your input everyone!
